### ISSUE TITLE
DeviceLost on a CanvasVirtualControl

### ISSUE THREAD

**Issue by ErwanJes** [Microsoft/Win2D#194]
I have a simple CanvasVirtualControl :
```xaml
<canvas:CanvasVirtualControl x:Name="canvas" />
```

```c#
canvas.RegionsInvalidated += OnRegionsInvalidated;
...
private void OnRegionsInvalidated(CanvasVirtualControl sender, CanvasRegionsInvalidatedEventArgs args)
{
      foreach (var region in args.InvalidatedRegions)
      {
        try
        {
          using (var session = sender.CreateDrawingSession(region))
          {
          }
        }
        catch (Exception ex)
        {
         sender.Device.IsDeviceLost(ex.HResult); // == True !
        }
    }
}

```

When the device wake up from sleep (or suspended mode), there is a call to `OnRegionsInvalidated` which crash in `sender.CreateDrawingSession(region)`.
Even though we catch the exception, there is a infinite loop there.

Is there a way to properly handle DeviceLost ?

PS : I read the doc http://microsoft.github.io/Win2D/html/HandlingDeviceLost.htm but did not find a solution.

**Comment by damyanp** [comment]
Win2D controls handle device lost exceptions thrown by event handlers that they invoke.  Event handler implementations do not need to try/catch for device lost.

So the correct implementation in this case should be:

```C#
private void OnRegionsInvalidated(CanvasVirtualControl sender, CanvasRegionsInvalidatedEventArgs args)
{
      foreach (var region in args.InvalidatedRegions)
      {
          using (var session = sender.CreateDrawingSession(region))
          {
          }
    }
}
```

Please let us know if you find that this does not work in some situation!

I notice that in HandlingDeviceLost.htm we list the CreateResources, Update and Draw event handlers.  We need to add RegionsInvalidated to this list.  I'll do that now.

**Comment by ErwanJes** [comment]
Thank you for you quick answer.

Unfortunately, I tried to let Win2D manage this exception (no try/catch in my code) but I have the same issue.

**Comment by damyanp** [comment]
Is this the same as issue #163?

**Comment by ErwanJes** [comment]
It looks like a lot with a few differences :
- I cannot reproduce the same bug with the Win2d samples, at least not at 100%
- It's a Universal App
- The exception is not handled, the application crashes.

**Comment by damyanp** [comment]
Ok.  Can you confirm that the exact code you have is as shown above https://github.com/Microsoft/Win2D/issues/194#issuecomment-161711963 and provide me with repro steps please?

Thanks.

**Comment by ErwanJes** [comment]
- Yes it is the exact code
- The reproduction is exactly the same as [#163](https://github.com/Microsoft/Win2D/issues/163#issuecomment-143881105) with my application

**Comment by damyanp** [comment]
Hmmm....I'm afraid I'm not seeing this.   Can you share the entire project with me somehow?  Since you're unable to repro this when running ExampleGallery it really does suggest that there's some difference between your project and my attempt to reproduce your project.

Can you think of any other steps I may be missing?  Are you running x86 or x64?  Debug or release?

Also just to confirm the version of Windows that you're running on, could you run the following from a command prompt:
````reg query "HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion" /v BuildLabEx````

**Comment by ErwanJes** [comment]
I will try to create a small sample to give it to you.

- Yes there may be differences for this bug but I did not notice them
- x64 / Debug
- `BuildLabEx    REG_SZ    10240.16590.amd64fre.th1_st1.151104-1714`

**Comment by Datoh** [comment]
I found the issue, we have a model notification and we call an invalidate:

**Comment by damyanp** [comment]
Hi - excellent!  

Invalidate() on CanvasVirtualControl should not throw device lost, but it looks like it is.  We'll look into addressing this soon.

In the meantime, I'd suggest:

1. Invalidate() is intended to be callable from any thread, so you don't need to dispatch back to the UI thread.
2. Add a try/catch around the call to Invalidate:

```C#
public void OnModelUpdated()
{
    try 
    { 
        Canvas.Invalidate(); 
    }
    catch (Exception e) where Canvas.Device.IsDeviceLost(e.ErrorCode)
    { 
        Canvas.Device.RaiseDeviceLost(); 
    }
}

Alternatively, if ReadyToDraw works well enough for you then you may want to stick with that.
```

**Comment by damyanp** [comment]
(Note my previous comment was incorrect about CanvasVirtualControl.Invalidate() being callable from any thread. )

**Comment by damyanp** [comment]
Actually, I think I spoke too soon.  I've not been able to contrive any situation where Invalidate() throws.  Before adding a fix for this I'd like to see the problem happen on a machine where I can determine the root cause.

So - if you could continue putting together a project that reproduces this then that would be most helpful.

Thanks!

**Comment by ErwanJes** [comment]
We check if `canvas.ReadyToDraw` before `canvas.Invalidate()`, seems to fix the crash.

But I could not find a small sample to reproduce the crash...

**Comment by damyanp** [comment]
Ok.  I think there's not much further we can do here, so I'm going to close this issue.  Feel free to reopen if you are able to get a repro you can share with us.

Thanks!

**Comment by ErwanJes** [comment]
To continue with this issue, sometimes after a resume, there is no callback `RegionsInvalidated` after a `canvas.Invalidate(rect)`.

I do not know if this is link to the previous problem but I think so.

With Win2D in Debug, `imageSource` is null, so nothing is fired.
```C++
IFACEMETHODIMP CanvasVirtualControl::InvalidateRegion(Rect region)
{
    return ExceptionBoundary(
        [&]
        {
            auto imageSource = GetCurrentRenderTarget()->Target;
            // imageSource = nullptr
            if (imageSource)
                ThrowIfFailed(imageSource->InvalidateRegion(region));
        });
}
```
I still cannot repro this in a small sample :/

**Comment by damyanp** [comment]
What do you see on screen when this happens?  I would hope that shortly after this happening the image source will be created and, since it is all invalid, RegionsInvalidate will be raised shortly afterwards to ensure that the visible region has something drawn in it.

**Comment by ErwanJes** [comment]
The canvas is empty, I don't get any RegionsInvalidated callback.

**Comment by damyanp** [comment]
Interesting.  Can you let us know when you've got a repro for this you can share with us?

Thanks!

**Comment by ErwanJes** [comment]
I finally was able to repro the issue from your SimpleSample.
https://github.com/ErwanJes/SimpleInking-Win2D

**Comment by damyanp** [comment]
What are the repro steps?

**Comment by ErwanJes** [comment]
Ah sorry :\
1) Draw some strokes
2) Suspend the application
3) Resume

4) Crash on CreateSession or canvas empty

**Comment by damyanp** [comment]
Ok...that's what I was trying, but I've not managed to reproduce.  

Is there anything else I should consider?  Should I be running on PC or phone?  x86?  x64?  Release? Debug? etc.

Should it repro with the debugger attached?

**Comment by Datoh** [comment]
Hi, I can reproduce on my device (Surface Pro 3).

1. Open the .sln from https://github.com/ErwanJes/SimpleInking-Win2D in VS2015
2. Run the project on a local machine (Debug/x64).
3. Draw some strokes.
4. Switched to VS, and in the lyfecycle events (Debug Location), pick "Suspend", then 2 seconds later, pick "Resume"
5. Go back to the sample, and then the canas is empty or the sample crash.

**Comment by damyanp** [comment]
(still no luck reproducing) have you seen this repro on a machine running the November update?  (10586)

**Comment by ErwanJes** [comment]
I can reproduce it every time (also on Surface Pro 3) with the november update.

**Comment by Datoh** [comment]
My last win10 update is the 1511 Buid 10586.29

**Comment by damyanp** [comment]
Would you be able to provide a recording of the repro so we can be sure we're doing exactly the same thing as you?  We're obviously missing something somewhere...

Thanks.

**Comment by Datoh** [comment]
Here is a screencast

![animation](https://cloud.githubusercontent.com/assets/2086938/11730232/f7181890-9f93-11e5-87cc-871b7bcdadd4.gif)

**Comment by shawnhar** [comment]
Odd...   I'm doing the exact same thing shown in this video but am still unable to repro any issue.

**Comment by ErwanJes** [comment]
argh :/
What device do you use ?

**Comment by damyanp** [comment]
We're trying to get hold of a Surface Pro 3 to try this on.  I looks like suspending the app on your device is causing the D3D device to be lost (you see the "Removing Device" message in the debug window).  We don't see this on desktop.  

We wouldn't normally expect a device lost as a result of suspending the app, so there's something interesting going on here.

**Comment by ErwanJes** [comment]
Yes there is a message regarding 3D3 device in the output.
Thanks

**Comment by damyanp** [comment]
You'll be pleased to hear that I can reproduce this on a Surface Pro 3!   It does seem that whenever I suspend an app (even one that just has a CanvasControl in it) the device is lost.

This only happens when using the hardware device - if you set ForceSoftwareRendering to True on the controls then it no longer repros.  This may be a way to allow you to validate your scenarios without running into this.

Of course, we'll continue to work to resolve the underlying issue here.

Thank you for your patience and persistence!

**Comment by ErwanJes** [comment]
Thank you for looking into this.

Is using `ForceSoftwareRendering` will impact the performances ?

**Comment by 13thsymphony** [comment]
Yes, this will prevent using the GPU hardware and use software emulation. It is fast enough to be quite usable generally, but will definitely be slower than using real GPU hardware.

**Comment by damyanp** [comment]
There's two problems at play here:  

One is a Win2D bug, which sent us all down the wrong path.  The issue here was that we weren't clearing the device lost error when Win2D handles it.  In WinRT there's a bit of per-thread state containing extended information about the current exception - this is what Win2D uses to put human-readable exception descriptions in the debugger.  When XAML sees an error it checks for this extended information, and if it is there it uses this to override whatever error its handling.  So when there's an error that doesn't set this extended state after the device lost error the device lost error ends up being reported - hiding the real issue.  (This will be fixed in the next Win2D release.)

So the second problem is a bug in your app that you're going to have to handle.  The problem is that any geometry that you store between draws gets invalidated when the device is lost.  When you draw geometry created with the old device this will fail.  Due to the batching / deferred nature of XAML and Win2D you normally only see this error in the UnhandledException handler.

If you enable the debug layer (set CanvasDevice.DebugLevel = CanvasDebugLevel.Information before creating any Win2D resources) then you'll get a debug break closer to where the error occurs.

**Comment by damyanp** [comment]
Attached image shows the error as reported when the debug layer (and native debugging) is enabled:

![image](https://cloud.githubusercontent.com/assets/8118402/11945173/b10f2598-a801-11e5-90e6-55cf3104bfbb.png)

**Comment by ErwanJes** [comment]
Thanks for your help.
I will try to fix it on our side.

**Comment by shawnhar** [comment]
Win2D 1.12.0 (released today) fixes the issue where we weren't clearing the device lost error after Win2D handles it.  With this change, you will now see the right exception message caused by the app logic error.

**Comment by ErwanJes** [comment]
We fix the crash caused by the factory released but we still have an issue.
Now the canvas is not redraw after the resume, sometimes a small part of the view, sometimes all the screen. It may be link to the use of a ScrollViewer. The issue happens randomly.

![animation](https://cloud.githubusercontent.com/assets/1706728/12644982/03107c4c-c5c7-11e5-93de-43f259d94240.gif)

A screenshot of the problem with the whole screen.
<img width="1077" alt="win2d_suspend" src="https://cloud.githubusercontent.com/assets/1706728/12645008/36c98cb8-c5c7-11e5-9cd6-4640bd45754a.png">


Here is a sample to reproduce the issue:
https://github.com/ErwanJes/SimpleInking-Win2D

**Comment by damyanp** [comment]
As a workaround, have you tried calling Invalidate() on the CanvasVirtualControl when the application is resumed?

**Comment by ErwanJes** [comment]
Thanks for your quick answer.
On the sample, I add a call to `canvas.Invalidate()` when the application is resumed. Most of the time, the strokes are shown, but the inking is very laggy with a lot of exceptions thrown:

```
D3D11: Removing Device.
D3D11 WARNING: ID3D11Device::RemoveDevice: Device removal has been triggered for the following reason (DXGI_ERROR_DRIVER_INTERNAL_ERROR: There is strong evidence that the driver has performed an undefined operation; but it may be because the application performed an illegal or undefined operation to begin with.). [ EXECUTION WARNING #379: DEVICE_REMOVAL_PROCESS_POSSIBLY_AT_FAULT]
Exception thrown at 0x00007FFA614D1F08 in SimpleSample.uap.exe: Microsoft C++ exception: _com_error at memory location 0x000000B5DD0FA628.
...
Exception thrown at 0x00007FFA614D1F08 in SimpleSample.uap.exe: Microsoft C++ exception: _com_error at memory location 0x000000B5DD0FDA80.
D2D DEBUG WARNING - A Draw call by a render target failed [8899000c]. Tags [0x0, 0x0].
```

**Comment by SebastianStehle** [comment]
@ErwanJes I have a similar problem (see referenced issue, https://github.com/Microsoft/Win2D/issues/300). Do you have a stable workaround?

**Comment by shawnhar** [comment]
I've been digging into this some more and am convinced that the remaining issues you are seeing (after Damyan's fix) are identical to #300.  I'm going to close this as a dupe and use #300 to track getting this fixed.
